### DISCUSSION TITLE
Is there some way to create a gatsby-like API that is available during build time?

### DISCUSSION THREAD

**Discussion by good-idea** [vercel/next.js#32380] [Help]
I have a project that generates 200+ pages, and within this page content there are a *lot* of instances where I need to query for relational data within my larger dataset.

Specifically, it's consuming data from Sanity, and I need to make a lot of expanded reference queries. (i.e. a page's rich text might link to another document, to get that linked document's URL I need to expand on the reference, then within the linked document, expand one or two more times to get it's parent docs, then finally use all of that information to generate the appropriate URL). Some pages might have dozens of referenced articles, and since each page (using `getStaticPaths`) runs in its own container, *every path* results in very large queries and long build times.

This made me think of how Gatsby works - it pulls *all* of your content from your data source and provides a single GraphQL api to get this information when you are building your pages. Since all of your content is fetched ahead of time, you don't need to keep hitting your data source's API during all of your builds -- all the data you need is available in memory.

Ideally, I'd like to be able to have some kind of server or container that was available to all page builds during build time. I've tried doing some hacky stuff by reading and writing to the filesystem, but I'm using vercel and this has been very buggy.

It would be great to be able to do something like this in my page builds:

```ts
export const getStaticProps ({ document }) => {
   const documentHref = myBuildServer.getDocumentHref(document._id)
   // etc
}
```

*OR*, less ideal but still doable, update my build command to be something like `yarn generateSiteJson && yarn next build`, then:

```ts
const siteJson = require('../generated/site.json')

export const getStaticProps ({ document }) => {
   const documentHref = siteJson.find(...)
   // etc
}
```

Does anyone have any solutions for something like this?
Are there any plans for an official solution?

**Comment by marceloverdijk** [comment]
@good-idea How did you implement this eventually?
